### PR TITLE
AI: reduce overhead of sample counter

### PR DESCRIPTION
**Pull request by booto**
This was created to address the problem with VC titles loading.

Needs testing.

### CODE REVIEW

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by t27duck** [comment]
Did some very light smokescreen testing with (either made sure the title screen came up or briefly started one level):

Super Mario Bros.
Kirby's Adventure
Super Mario World
Donkey Kong Country
Super Mario 64
Mario Kart 64
F-Zero X
Sonic and Knuckles
Sonic the Hedgehog 3
Sonic Spinball

All seem to boot up and (eventually) sound output catches up for those who don't output audio immediately. F-zero X might have a ~0.2 second delay on the boost sound effects... but that might be my imagination.

Started up Sonic Colors, DKCR, and Excite Truck successfully too (a just in case check).

All my games are NA and I'm using the Windows build for this branch.

**Comment by booto** [comment]
For which titles is this "sound output catchup" thing most apparent?

**Comment by phire** [comment]
Does the sound catch up effect go away if you use a backend other than OpenAI?

**Comment by Linktothepast** [comment]
All the n64 games and the neo geo games that i tested worked fine.

**Comment by t27duck** [comment]
Apologies... I forgot to run the build with a portable.txt file so it was using all of my custom settings and stuff when I went through it the first time.  :disappointed: 

Using default settings sounds seems to sync up at the start correctly with the games. Carry on.

**Comment by phire** [comment]
LGTM

**Comment by mbc07** [comment]
Not sure if it's related to this PR, but Epic Mickey/Epic Mickey 2 both start completely silent when using OpenAL and remain like that for at least 15-20 seconds, when the sound start playing veeeery fast until it catches up with current output. It's the same behavior in master and this PR apparently has no effect at all...

**Comment by phire** [comment]
[Issue 7955](https://bugs.dolphin-emu.org/issues/7955)

**Comment by booto** [comment]
@mbc07 I don't known anything about the problem your describe, and from a cursory glance there doesn't seem to be an issue for it already.

Could you please [submit one](https://bugs.dolphin-emu.org/projects/emulator/issues/new)?

**Comment by mbc07** [comment]
@booto did a quick look at [issue 7955](https://bugs.dolphin-emu.org/issues/7955) and it seems to be the same problem...

File: Source/Core/Core/HW/AudioInterface.cpp [hunk 1]
```diff
@@ -308,16 +307,19 @@ void Update(u64 userdata, int cyclesLate)
...
-u64 GetAIPeriod()
+int GetAIPeriod()
 {
-	u64 period = g_CPUCyclesPerSample * m_InterruptTiming;
-	if (period == 0)
-		period = 32000 * g_CPUCyclesPerSample;
-	return period;
+	u64 period = g_CPUCyclesPerSample * (m_InterruptTiming-m_SampleCounter);
+	u64 s_period = g_CPUCyclesPerSample * g_AISSampleRate;
+	if ((period == 0) || (s_period < period))
```
Comment (degasus) [inline Source/Core/Core/HW/AudioInterface.cpp]:
```
if (period == 0)
  return s_period;
return std::min(period, s_period)
```

**Comment by JMC47** [comment]
Requesting merge if there are no more complaints.

**Comment by degasus** [comment]
The changes looks good to me, but why do you always use "GetAIPeriod() / 2"?
Why is this divde required at all?

**Comment by booto** [comment]
It was in the original code, and I was trying to change as little as possible so as to not inadvertently (re)introduce bugs. The div2 can be moved to GetAIPeriod as it's always applied to the result.

I'll try removing the div2 and test as many things as possible (esp. titles that have traditionally been sensitive to the sample counter). If there are any problems found, I'll revert back, but move the div2 into the GetAIPeriod function.

**Comment by degasus** [comment]
LGTM

**Pull request merged**